plugin/trino-faker/src/main/java/io/trino/plugin/faker/FakerPageSourceProvider.java

```diff
@@ -0,0 +1,785 @@
+            List<ColumnHandle> columns,
+            DynamicFilter dynamicFilter)
+    {
+        List<FakerColumnHandle> handles = columns
+                .stream()
+                .map(c -> (FakerColumnHandle) c)
+                .collect(toList());
+
+        FakerTableHandle fakerTable = (FakerTableHandle) table;
+
+        return new PageSource(faker, random, handles, fakerTable);
+    }
+
+    private static class PageSource
+            implements ConnectorPageSource
+    {
+        private static final int ROWS_PER_REQUEST = 4096;
+
+        private final Random random;
+        private final Faker faker;
+        private final FakerTableHandle table;
+        private final List<BiConsumer<BlockBuilder, Long>> generators;
+        private long completedBytes;
+        private long completedRows;
+
+        private final PageBuilder pageBuilder;
+
+        private boolean closed;
+
+        private PageSource(Faker faker, Random random, List<FakerColumnHandle> columns, FakerTableHandle table)
+        {
+            this.faker = requireNonNull(faker, "faker is null");
+            this.random = requireNonNull(random, "random is null");
+            this.table = requireNonNull(table, "table is null");
+            List<Type> types = requireNonNull(columns, "columns is null")
+                    .stream()
+                    .map(FakerColumnHandle::getType)
+                    .collect(toList());
+            this.generators = columns
+                    .stream()
+                    .map(column -> constraintedValueGenerator(
+                            column,
+                            table.getConstraint().getDomains().get().getOrDefault(column, Domain.all(column.getType()))))
+                    .collect(toList());
+            // TODO create generators for every column
+            this.pageBuilder = new PageBuilder(types);
+        }
+
+        @Override
+        public long getCompletedBytes()
+        {
+            return completedBytes;
+        }
+
+        @Override
+        public long getReadTimeNanos()
+        {
+            return 0;
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            return closed && pageBuilder.isEmpty();
+        }
+
+        @Override
+        public Page getNextPage()
+        {
+            if (!closed) {
+                int positions = (int) Math.min(table.getLimit() - completedRows, ROWS_PER_REQUEST);
+                if (positions <= 0) {
+                    closed = true;
+                }
+                else {
+                    pageBuilder.declarePositions(positions);
+                    for (int column = 0; column < generators.size(); column++) {
+                        BlockBuilder output = pageBuilder.getBlockBuilder(column);
+                        for (int i = 0; i < positions; i++) {
+                            generators.get(column).accept(output, completedRows + i);
+                        }
```

Comment:
You should loop over the columns and append N positions at a time to each column, that is more efficient than looping over columns for each position

Reply:
I'm a bit confused, isn't this already the case, since this is an inner loop? I just extracted `generators.get(column)` outside this loop.